lib/vector-core/src/schema/definition.rs

```diff
@@ -108,27 +141,36 @@ impl Definition {
         kind: Kind,
         meaning: Option<&str>,
     ) -> Self {
-        let mut path = path.into();
+        let path = path.into();
         let meaning = meaning.map(ToOwned::to_owned);
 
-        match path.get(0) {
-            None => panic!("must not be a root path"),
-            Some(segment) if segment.is_index() => panic!("must not start with an index"),
-            _ => {}
-        };
-
-        let collection = kind
-            .nest_at_path(
-                &path.to_lookup(),
-                nest::Strategy {
-                    coalesced_path: nest::CoalescedPath::Reject,
-                },
-            )
-            .expect("non-coalesced path used")
-            .into_object()
-            .expect("always object");
+        if !path.is_root() {
+            if kind.contains_null() {
+                // field is optional, so don't coerce to an object, but still make sure it _can_ be an object
```

Comment:
See https://github.com/vectordotdev/vector/pull/13103 which originally made this change. I'm glad to talk about this more though. I understand your concern here, but the previous approach did the same thing, it just did more work and made it more complicated to understand. Accessing a non-existent field today returns null, so we can't get away from this concept unless that is changed. If we want to separate this, we need to introduce the `undefined` type, which might be a pretty large change (to make it user friendly)

Reply:
> I assume that's because it wanted to keep the distinction between undefined and null

It _wanted_ to, but it actually didn't. The problem is that the type system doesn't actually support the distinction between `null` and `undefined`. The `definition` was basically "pretending" to support this difference, but then when a VRL program was compiled, the optionality was _then_ baked into the type as `null`. After remap, you now have a definition that contains a empty list of `optional` fields, even though some of those values are actually optional (with a type that contains `null`). You can have other components that now take this definition as an input, and now the `optional` fields is lying since it's empty. In order to repopulate that empty list, you would need to rely on the type system to tell you which ones are "optional", but if the type system can tell you that, there's no reason to keep a separate list.

> For example, we also have the `exists` function in VRL

It's possible to _differentiate_ between "null" and "undefined" at _runtime_ today, as the `exists` function demonstrates, but the type system doesn't currently support encoding this, and there also isn't an `undefined` runtime value that exists yet that could be returned, so this concept can't be fully implemented without more changes.

Reply:
Ah yes, I should have clarified what I meant. I am aware that the type system in **VRL** doesn't support the distinction between `null` and `undefined`, but what I am getting at, is that the schema system in Vector _did_ support that, and since not every topology includes a `remap` transform, those invariants are upheld and available for sinks to validate against.

Having said all that, since the concept of optionality was removed in a previous PR, and we don't _currently_ have a source/sink combination that requires this feature, I'm okay leaving it out for now, as long as we make sure there's a follow-up issue to track adding support back, at which point we can bake it straight into the `Kind` type, which then allows easier integration back into VRL as well.

Reply:
> that the schema system in Vector did support that, and since not every topology includes a remap transform, those invariants are upheld and available for sinks to validate against.

Yes, this might work if remap isn't used, but as soon as a remap is added it would stop working, so we would need to add support to the type system for it to be reliably used in Vector. 
I added an issue (https://github.com/vectordotdev/vector/issues/13459) to track this work.